### DISCUSSION TITLE
Possible symbolic matrix element?

### DISCUSSION THREAD

**Discussion by eliotwrobson** [PrairieLearn/PrairieLearn#7027] [Dev]
After some discussion on slack, it seems like there may be some situations where it might be useful to have students input a matrix of symbolic answers (as opposed to numerical ones). Creating an element for doing this would be pretty straightforward, since we can just use a mustache template similar to that of `pl-matrix-input` and shared backend sympy code.

Is this a common use case? What do others think?

Some similar use cases to https://github.com/PrairieLearn/PrairieLearn/pull/6139

After some thought, I think the steps towards this are as follows (each a separate PR):
- [ ] Finish outstanding refactors for backend sympy code: https://github.com/PrairieLearn/PrairieLearn/pull/7020
- [ ] Refactor and clean up `pl-matrix-component-input`, since it will share backend with this new element. https://github.com/PrairieLearn/PrairieLearn/pull/7198
- [ ] Write the new `pl-matrix-symbolic-input` using common code from the previous two items.

**Comment by echuber2** [comment]
I implemented one of these for Prof. Hoiem's CS 445 way back, which may be useful for reference but isn't the ideal way to do it:

https://github.com/PrairieLearn/PrairieLearn/pull/2392

I had to do some hackery to get it to accept inputs with a decimal point as desired (the default symbolic input at the time rejected floats. Maybe it still does?). It would be preferable to have the matrix element as a template that embeds the upstream symbolic input instead of forking it.

**Reply by eliotwrobson** [reply]
Yes, the current symbolic input still rejects floats. I think there might be issues where floating point numbers may not be tested correctly for equality? If we want to support that though, that can be baked into the symbolic input code as a separate feature.

But yes, I think that the existing symbolic input code has been written to facilitate being a shared backend for different types of elements. So I think the code for this will mostly be building out the display with a few different options.

**Reply by echuber2** [reply]
There are issues with equality testing, yes. The use case when I built that version was only concerned about simple cases.

**Reply by echuber2** [reply]
Just to clarify, I think that float issue was the only reason I had to fork core PL code and the symbolic input element itself to make this work. Apart from that, it was just building out the display. So that part of what I did may be useful.

**Reply by eliotwrobson** [reply]
That makes sense, I think your code will be a nice starting point if we go forward with this.
